README
codebird-js
Copyright (C) 2011-2013 J.M. <me@mynetx.net>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


This is the JScript version of the Codebird library.
It was originally introduced in the Tweet2PSM Messenger Plus! script.
When using in other JScript environments (such as node.js),
you should modify the way how calls to external servers are done.



CHANGELOG
codebird-js - changelog
=======================

3.0.0 (not yet released)
+ #106 Add error callback
+ #96 Add logout method
+ #98 Support promises as alternative to callbacks
- Drop support for undocumented API methods
+ Add security check hasOwnProperty
+ #110 Add support for Collections API
+ Transform codebase to EcmaScript 7
+ #25 Add unit testing suite
+ Add POST statuses/unretweet/:id
+ #116 Update GET statuses/oembed with new endpoint
+ Support POST media/upload with media_data base64
- Fix check for require for webpack compatibility
+ Add Account Activity API methods
+ Add new Direct Messages API
- Remove sunsetting Direct Messages API
- Remove contributor API methods
- Remove deprecated statuses/update_with_media method
- Remove deprecated statuses/update_profile_background_image method
+ Detect XML error replies (re-implemented detection)
+ Add oauth/invalidate_token method

2.6.0 (2015-04-08)
+ Allow to get the supported API methods as array
- #79 Use POST for users/lookup and statuses/lookup, params may get too long for GET
- #81 Special oauth_ methods: Fix force_login parameter

2.5.0 (2014-06-23)
+ #24 Add README section about xAuth
+ #46 Don't force to send base64 media uploads via proxy
+ #23 Support internal API methods
+ #50 Current rate sent back with each call
+ Support custom timelines beta API
- Remove POST geo/place API method, deprecated
+ #52 Don't use proxy for non-browser environments
- #49 Places & Geo API call issue
+ #53 Support AMD module definition syntax
+ #54 Integrate sha1 hash library
+ Add NPM package.json
- #55 Typo in _clone function
- #57 Fix XHR in Safari
+ Add new API methods
+ #60 Multiple media per tweet
+ #66 Support Titanium
- #71 Return rate-limiting info only if contained in response
- #72 Don't require consumer key if bearer token is already known

2.4.3 (2013-07-24)
- #31 HTTP error on multiple oauth_requestToken calls
- #30 HTML-escaping typo in readme code sample

2.4.2 (2013-07-09)
- #21 Typo extra bracket in readme sample code
- #27 Array bad behavior in iOS

2.4.1 (2013-06-23)
+ Pass JSHint validation
+ #15 Allow to assign custom proxy address
- #19 API parameters with escaped characters fail in IE7-9

2.4.0 (2013-06-15)
+ Support multipart file uploads
- Drop unnecessary media format settings ported from PHP getimagesize
- Don't unintendedly overwrite global "url" variable
+ Add contributing guidelines

2.3.3 (2013-05-31)
- oauth2_token doesn't respect _useProxy CORS setting
- #10 setBearerToken was not public

2.3.2 (2013-05-21)
- Sorting the OAuth parameter list in IE incorrectly
- Undefined function c for debug logging
+ Correctly detect templated methods and insert their parameters

2.3.1 (2013-04-26)
- Wrong OAuth signature generation for POST requests with oauth_callback
- Wrong _ksort implementation

2.3.0 (2013-04-26)
- Bug in _detectMultipart for statuses/update calls
+ Add Array.indexOf polyfill
+ Circumvent same-origin policy with CORS
+ Make oauth_authenticate and oauth_authorize asynchronous
+ Send GET requests via JSONP for IE7-9 (don't support CORS)

2.2.3 (2012-12-03)
+ Detect HTTP method properly when accessing multi HTTP method endpoint (account/settings)
+ Add friends/list, followers/list API methods
+ Move users/recommendations to GET list

2.2.2 (2012-10-19)
+ Update methods list for API 1.1
+ Add profile banner methods
- Drop separate media upload method
- Remove OAuth token message. 1.1 is always signed.
+ Update endpoints to 1.1 endpoints, add support for old endpoints

2.2.1 (2012-09-17)
+ Update to Twitter API v1.1

2.2.0 (2012-09-17)
+ Port Codebird from PHP to JavaScript


test/media_tests.js
const tape = require("tape"),
  _test = require("tape-promise"),
  test = _test(tape), // decorate tape
  CodebirdT = require("./codebirdt");

function getCB() {
  var cb = new CodebirdT.default;
  cb.setConsumerKey("123", "456");

  return cb;
}

test("Tests media/upload base64", function (t) {
  const cb = getCB();
  t.plan(1);

  cb.setToken("123", "456");
  return cb.__call(
    "media_upload",
    {"media_data": "R0lGODlhPQBEAPeoAJosM//AwO/AwHVYZ/z595kzAP/s7P+goOXMv8+fhw/v739/f+8PD98fH/8mJl+fn/9ZWb8/PzWlwv///6wWGbImAPgTEMImIN9gUFCEm/gDALULDN8PAD6atYdCTX9gUNKlj8wZAKUsAOzZz+UMAOsJAP/Z2ccMDA8PD/95eX5NWvsJCOVNQPtfX/8zM8+QePLl38MGBr8JCP+zs9myn/8GBqwpAP/GxgwJCPny78lzYLgjAJ8vAP9fX/+MjMUcAN8zM/9wcM8ZGcATEL+QePdZWf/29uc/P9cmJu9MTDImIN+/r7+/vz8/P8VNQGNugV8AAF9fX8swMNgTAFlDOICAgPNSUnNWSMQ5MBAQEJE3QPIGAM9AQMqGcG9vb6MhJsEdGM8vLx8fH98AANIWAMuQeL8fABkTEPPQ0OM5OSYdGFl5jo+Pj/+pqcsTE78wMFNGQLYmID4dGPvd3UBAQJmTkP+8vH9QUK+vr8ZWSHpzcJMmILdwcLOGcHRQUHxwcK9PT9DQ0O/v70w5MLypoG8wKOuwsP/g4P/Q0IcwKEswKMl8aJ9fX2xjdOtGRs/Pz+Dg4GImIP8gIH0sKEAwKKmTiKZ8aB/f39Wsl+LFt8dgUE9PT5x5aHBwcP+AgP+WltdgYMyZfyywz78AAAAAAAD///8AAP9mZv///wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAEAAKgALAAAAAA9AEQAAAj/AFEJHEiwoMGDCBMqXMiwocAbBww4nEhxoYkUpzJGrMixogkfGUNqlNixJEIDB0SqHGmyJSojM1bKZOmyop0gM3Oe2liTISKMOoPy7GnwY9CjIYcSRYm0aVKSLmE6nfq05QycVLPuhDrxBlCtYJUqNAq2bNWEBj6ZXRuyxZyDRtqwnXvkhACDV+euTeJm1Ki7A73qNWtFiF+/gA95Gly2CJLDhwEHMOUAAuOpLYDEgBxZ4GRTlC1fDnpkM+fOqD6DDj1aZpITp0dtGCDhr+fVuCu3zlg49ijaokTZTo27uG7Gjn2P+hI8+PDPERoUB318bWbfAJ5sUNFcuGRTYUqV/3ogfXp1rWlMc6awJjiAAd2fm4ogXjz56aypOoIde4OE5u/F9x199dlXnnGiHZWEYbGpsAEA3QXYnHwEFliKAgswgJ8LPeiUXGwedCAKABACCN+EA1pYIIYaFlcDhytd51sGAJbo3onOpajiihlO92KHGaUXGwWjUBChjSPiWJuOO/LYIm4v1tXfE6J4gCSJEZ7YgRYUNrkji9P55sF/ogxw5ZkSqIDaZBV6aSGYq/lGZplndkckZ98xoICbTcIJGQAZcNmdmUc210hs35nCyJ58fgmIKX5RQGOZowxaZwYA+JaoKQwswGijBV4C6SiTUmpphMspJx9unX4KaimjDv9aaXOEBteBqmuuxgEHoLX6Kqx+yXqqBANsgCtit4FWQAEkrNbpq7HSOmtwag5w57GrmlJBASEU18ADjUYb3ADTinIttsgSB1oJFfA63bduimuqKB1keqwUhoCSK374wbujvOSu4QG6UvxBRydcpKsav++Ca6G8A6Pr1x2kVMyHwsVxUALDq/krnrhPSOzXG1lUTIoffqGR7Goi2MAxbv6O2kEG56I7CSlRsEFKFVyovDJoIRTg7sugNRDGqCJzJgcKE0ywc0ELm6KBCCJo8DIPFeCWNGcyqNFE06ToAfV0HBRgxsvLThHn1oddQMrXj5DyAQgjEHSAJMWZwS3HPxT/QMbabI/iBCliMLEJKX2EEkomBAUCxRi42VDADxyTYDVogV+wSChqmKxEKCDAYFDFj4OmwbY7bDGdBhtrnTQYOigeChUmc1K3QTnAUfEgGFgAWt88hKA6aCRIXhxnQ1yg3BCayK44EWdkUQcBByEQChFXfCB776aQsG0BIlQgQgE8qO26X1h8cEUep8ngRBnOy74E9QgRgEAC8SvOfQkh7FDBDmS43PmGoIiKUUEGkMEC/PJHgxw0xH74yx/3XnaYRJgMB8obxQW6kL9QYEJ0FIFgByfIL7/IQAlvQwEpnAC7DtLNJCKUoO/w45c44GwCXiAFB/OXAATQryUxdN4LfFiwgjCNYg+kYMIEFkCKDs6PKAIJouyGWMS1FSKJOMRB/BoIxYJIUXFUxNwoIkEKPAgCBZSQHQ1A2EWDfDEUVLyADj5AChSIQW6gu10bE/JG2VnCZGfo4R4d0sdQoBAHhPjhIB94v/wRoRKQWGRHgrhGSQJxCS+0pCZbEhAAOw=="}
  ).then(function (reply, rate, err) {
    t.deepEqual(
      reply,
      {
        rate: {
          limit: null, remaining: null, reset: null
        },
        reply: {
          errors: [
            { code: 89, message: "Invalid or expired token." }
          ],
          httpstatus: 200
        }
      }
    )
  });
});


test/oauth_tests.js
require("events").EventEmitter.defaultMaxListeners = 15;

const tape = require("tape"),
  _test = require("tape-promise"),
  test = _test(tape), // decorate tape
  CodebirdM = require("./codebirdm");

function getCB() {
  var cb = new CodebirdM.default();
  cb.setConsumerKey("123", "456");

  return cb;
}

test("Tests oauth_authenticate Promise", function(t) {
  const cb = getCB();
  t.plan(1);

  cb.setToken("123", "456");
  return cb.oauth_authenticate().then(function(a) {
    t.deepEqual(a, {
      reply: "https://api.twitter.com/oauth/authenticate?oauth_token=123"
    });
  });
});

test("Tests oauth_authenticate callback", function(t) {
  const cb = getCB();
  t.plan(4);

  cb.setToken("123", "456");
  cb.oauth_authenticate({}, function(a) {
    t.equal(a, "https://api.twitter.com/oauth/authenticate?oauth_token=123");
  });
  cb.oauth_authenticate({ force_login: true }, function(a) {
    t.equal(
      a,
      "https://api.twitter.com/oauth/authenticate?oauth_token=123&force_login=1"
    );
  });
  cb.oauth_authenticate(
    { force_login: true, screen_name: "TwitterAPI" },
    function(a) {
      t.equal(
        a,
        "https://api.twitter.com/oauth/authenticate?oauth_token=123&force_login=1&screen_name=TwitterAPI"
      );
    }
  );
  cb.oauth_authenticate({ screen_name: "TwitterAPI" }, function(a) {
    t.equal(
      a,
      "https://api.twitter.com/oauth/authenticate?oauth_token=123&screen_name=TwitterAPI"
    );
  });
});

test("Tests oauth_authorize callback", function(t) {
  const cb = getCB();
  t.plan(4);

  cb.setToken("123", "456");
  cb.oauth_authorize({}, function(a) {
    t.equal(a, "https://api.twitter.com/oauth/authorize?oauth_token=123");
  });
  cb.oauth_authorize({ force_login: true }, function(a) {
    t.equal(
      a,
      "https://api.twitter.com/oauth/authorize?oauth_token=123&force_login=1"
    );
  });
  cb.oauth_authorize({ force_login: true, screen_name: "TwitterAPI" }, function(
    a
  ) {
    t.equal(
      a,
      "https://api.twitter.com/oauth/authorize?oauth_token=123&force_login=1&screen_name=TwitterAPI"
    );
  });
  cb.oauth_authorize({ screen_name: "TwitterAPI" }, function(a) {
    t.equal(
      a,
      "https://api.twitter.com/oauth/authorize?oauth_token=123&screen_name=TwitterAPI"
    );
  });
});

test("Tests oauth2_token", function(t) {
  const cb = getCB();
  t.plan(1);

  cb.oauth2_token(function(a) {
    t.deepEqual(a, {
      token_type: "bearer",
      access_token: "VqiO0n2HrKE",
      httpstatus: 200
    });
  });
});

test("Tests signing of boolean parameters", function(t) {
  const cb = getCB();

  t.equal(
    cb.call("_getSignature", ["GET", "friends/ids", ["stringify_ids"], [true]]),
    "OFNuMTEnE82pfI0cAdJPgtO4xzY="
  );

  t.end();
});


test/codebirdm.es7.js
import CodebirdT from "./codebirdt";

/**
 * A Twitter library in JavaScript
 *
 * @package   codebird-test
 * @version   3.0.0-dev
 * @author    Jublo Limited <support@jublo.net>
 * @copyright 2010-2018 Jublo Limited <support@jublo.net>
 * @license   http://opensource.org/licenses/GPL-3.0 GNU Public License 3.0
 * @link      https://github.com/jublonet/codebird-php
 */

/**
 * A Twitter library in JavaScript
 *
 * @package codebird-test
 */
export default class CodebirdM extends CodebirdT {
  constructor() {
    super();
    /**
     * Mock API replies
     */
    this._mock_replies = {
      default: {
        httpstatus: 404,
        reply:
          '{"errors":[{"message":"Sorry, that page does not exist","code":34}]}'
      },
      "GET https://api.twitter.com/1.1/users/show.json?screen_name=TwitterAPI": {
        httpstatus: 200,
        reply:
          '{"id":6253282,"id_str":"6253282","name":"Twitter API","screen_name":"twitterapi","location":"San Francisco, CA","profile_location":null,"description":"The Real Twitter API. I tweet about API changes, service issues and happily answer questions about Twitter and our API. Don\'t get an answer? It\'s on my website.","url":"http://t.co/78pYTvWfJd","entities":{"url":{"urls":[{"url":"http://t.co/78pYTvWfJd","expanded_url":"http://dev.twitter.com","display_url":"dev.twitter.com","indices":[0,22]}]},"description":{"urls":[]}},"protected":false,"followers_count":4993679,"friends_count":48,"listed_count":13001,"created_at":"Wed May 23 06:01:13 +0000 2007","favourites_count":27,"utc_offset":-28800,"time_zone":"Pacific Time (US & Canada)","geo_enabled":true,"verified":true,"statuses_count":3553,"lang":"en","status":{"created_at":"Tue Nov 24 08:56:07 +0000 2015","id":669077021138493440,"id_str":"669077021138493440","text":"Additional 64-bit entity ID migration coming in Feb 2016 https://t.co/eQIGvw1rsJ","source":"\u003ca href=\\"https://about.twitter.com/products/tweetdeck\\" rel=\\"nofollow\\"\u003eTweetDeck\u003c/a\u003e","truncated":false,"in_reply_to_status_id":null,"in_reply_to_status_id_str":null,"in_reply_to_user_id":null,"in_reply_to_user_id_str":null,"in_reply_to_screen_name":null,"geo":null,"coordinates":null,"place":null,"contributors":null,"retweet_count":67,"favorite_count":79,"entities":{"hashtags":[],"symbols":[],"user_mentions":[],"urls":[{"url":"https://t.co/eQIGvw1rsJ","expanded_url":"https://twittercommunity.com/t/migration-of-twitter-core-entities-to-64-bit-ids/56881","display_url":"twittercommunity.com/t/migration-of\u2026","indices":[57,80]}]},"favorited":false,"retweeted":false,"possibly_sensitive":false,"lang":"en"},"contributors_enabled":false,"is_translator":false,"is_translation_enabled":false,"profile_background_color":"C0DEED","profile_background_image_url":"http://pbs.twimg.com/profile_background_images/656927849/miyt9dpjz77sc0w3d4vj.png","profile_background_image_url_https":"https://pbs.twimg.com/profile_background_images/656927849/miyt9dpjz77sc0w3d4vj.png","profile_background_tile":true,"profile_image_url":"http://pbs.twimg.com/profile_images/2284174872/7df3h38zabcvjylnyfe3_normal.png","profile_image_url_https":"https://pbs.twimg.com/profile_images/2284174872/7df3h38zabcvjylnyfe3_normal.png","profile_banner_url":"https://pbs.twimg.com/profile_banners/6253282/1431474710","profile_link_color":"0084B4","profile_sidebar_border_color":"C0DEED","profile_sidebar_fill_color":"DDEEF6","profile_text_color":"333333","profile_use_background_image":true,"has_extended_profile":false,"default_profile":false,"default_profile_image":false,"following":true,"follow_request_sent":false,"notifications":false}'
      },
      "POST https://api.twitter.com/oauth2/token": {
        httpstatus: 200,
        reply: '{"token_type":"bearer","access_token":"VqiO0n2HrKE"}'
      },
      "POST https://api.twitter.com/oauth/request_token": {
        httpstatus: 401,
        reply:
          '<?xml version="1.0" encoding="UTF-8"?><errors><error code="415">Callback URL not approved for this client application. Approved callback URLs can be adjusted in your application settings</error></errors>'
      }
    };

    this.xml = {
      readyState: 4,
      open: (httpmethod, url) => {
        this.xml.httpmethod = httpmethod;
        this.xml.url = url;
        const key = `${httpmethod} ${url}`;
        if (this._mock_replies.hasOwnProperty(key)) {
          this.xml.status = this._mock_replies[key].httpstatus;
          this.xml.responseText = this._mock_replies[key].reply;
        } else {
          this.xml.status = this._mock_replies.default.httpstatus;
          this.xml.responseText = this._mock_replies.default.reply;
        }
      },
      setRequestHeader: () => true,
      onreadystatechange: () => false,
      send: function() {
        setTimeout(this.onreadystatechange, 200);
      }
    };
  }

  _getXmlRequestObject() {
    return this.xml;
  }
}


sha1.js
/*
 * A JavaScript implementation of the Secure Hash Algorithm, SHA-1, as defined
 * in FIPS PUB 180-1
 * Version 2.1 Copyright Paul Johnston 2000 - 2002.
 * Other contributors: Greg Holt, Andrew Kepert, Ydnar, Lostinet
 * Distributed under the BSD License
 * See http://pajhome.org.uk/crypt/md5 for details.
 */

var hexcase = 0; 
var b64pad  = ""; 
var chrsz   = 8; 

function hex_sha1(s){return binb2hex(core_sha1(str2binb(s),s.length * chrsz));}
function b64_sha1(s){return binb2b64(core_sha1(str2binb(s),s.length * chrsz));}
function str_sha1(s){return binb2str(core_sha1(str2binb(s),s.length * chrsz));}
function hex_hmac_sha1(key, data){ return binb2hex(core_hmac_sha1(key, data));}
function b64_hmac_sha1(key, data){ return binb2b64(core_hmac_sha1(key, data));}
function str_hmac_sha1(key, data){ return binb2str(core_hmac_sha1(key, data));}

function sha1_vm_test()
{
  return hex_sha1("abc") == "a9993e364706816aba3e25717850c26c9cd0d89d";
}

function core_sha1(x, len)
{
  x[len >> 5] |= 0x80 << (24 - len % 32);
  x[((len + 64 >> 9) << 4) + 15] = len;

  var w = Array(80);
  var a =  1732584193;
  var b = -271733879;
  var c = -1732584194;
  var d =  271733878;
  var e = -1009589776;

  for(var i = 0; i < x.length; i += 16)
  {
    var olda = a;
    var oldb = b;
    var oldc = c;
    var oldd = d;
    var olde = e;

    for(var j = 0; j < 80; j++)
    {
      if(j < 16) w[j] = x[i + j];
      else w[j] = rol(w[j-3] ^ w[j-8] ^ w[j-14] ^ w[j-16], 1);
      var t = safe_add(safe_add(rol(a, 5), sha1_ft(j, b, c, d)), 
                       safe_add(safe_add(e, w[j]), sha1_kt(j)));
      e = d;
      d = c;
      c = rol(b, 30);
      b = a;
      a = t;
    }

    a = safe_add(a, olda);
    b = safe_add(b, oldb);
    c = safe_add(c, oldc);
    d = safe_add(d, oldd);
    e = safe_add(e, olde);
  }
  return Array(a, b, c, d, e);
  
}

function sha1_ft(t, b, c, d)
{
  if(t < 20) return (b & c) | ((~b) & d);
  if(t < 40) return b ^ c ^ d;
  if(t < 60) return (b & c) | (b & d) | (c & d);
  return b ^ c ^ d;
}

function sha1_kt(t)
{
  return (t < 20) ?  1518500249 : (t < 40) ?  1859775393 :
         (t < 60) ? -1894007588 : -899497514;
}  

function core_hmac_sha1(key, data)
{
  var bkey = str2binb(key);
  if(bkey.length > 16) bkey = core_sha1(bkey, key.length * chrsz);

  var ipad = Array(16), opad = Array(16);
  for(var i = 0; i < 16; i++) 
  {
    ipad[i] = bkey[i] ^ 0x36363636;
    opad[i] = bkey[i] ^ 0x5C5C5C5C;
  }

  var hash = core_sha1(ipad.concat(str2binb(data)), 512 + data.length * chrsz);
  return core_sha1(opad.concat(hash), 512 + 160);
}

function safe_add(x, y)
{
  var lsw = (x & 0xFFFF) + (y & 0xFFFF);
  var msw = (x >> 16) + (y >> 16) + (lsw >> 16);
  return (msw << 16) | (lsw & 0xFFFF);
}

function rol(num, cnt)
{
  return (num << cnt) | (num >>> (32 - cnt));
}

function str2binb(str)
{
  var bin = Array();
  var mask = (1 << chrsz) - 1;
  for(var i = 0; i < str.length * chrsz; i += chrsz)
    bin[i>>5] |= (str.charCodeAt(i / chrsz) & mask) << (24 - i%32);
  return bin;
}

function binb2str(bin)
{
  var str = "";
  var mask = (1 << chrsz) - 1;
  for(var i = 0; i < bin.length * 32; i += chrsz)
    str += String.fromCharCode((bin[i>>5] >>> (24 - i%32)) & mask);
  return str;
}

function binb2hex(binarray)
{
  var hex_tab = hexcase ? "0123456789ABCDEF" : "0123456789abcdef";
  var str = "";
  for(var i = 0; i < binarray.length * 4; i++)
  {
    str += hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8+4)) & 0xF) +
           hex_tab.charAt((binarray[i>>2] >> ((3 - i%4)*8  )) & 0xF);
  }
  return str;
}

function binb2b64(binarray)
{
  var tab = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
  var str = "";
  for(var i = 0; i < binarray.length * 4; i += 3)
  {
    var triplet = (((binarray[i   >> 2] >> 8 * (3 -  i   %4)) & 0xFF) << 16)
                | (((binarray[i+1 >> 2] >> 8 * (3 - (i+1)%4)) & 0xFF) << 8 )
                |  ((binarray[i+2 >> 2] >> 8 * (3 - (i+2)%4)) & 0xFF);
    for(var j = 0; j < 4; j++)
    {
      if(i * 8 + j * 6 > binarray.length * 32) str += b64pad;
      else str += tab.charAt((triplet >> 6*(3-j)) & 0x3F);
    }
  }
  return str;
}